src/net.cpp

```diff
@@ -1858,13 +1858,14 @@ void CConnman::ThreadMnbRequestConnections()
         std::pair<CService, std::set<uint256> > p = mnodeman.PopScheduledMnbRequestConnection();
         if(p.first == CService() || p.second.empty()) continue;
 
-        CNode* pnode = NULL;
-        {
-            LOCK2(cs_main, cs_vNodes);
-            pnode = ConnectNode(CAddress(p.first, NODE_NETWORK), NULL, true);
```

Comment:
Not part of this PR, but as far as I understand we leak node refs here. ConnectNode does an AddRef here no matter if it's a new connection or an already existing one. This means that the first call to ConnectNode is fine here, as it results in ref=1, basically meaning that CConman exclusively owns the node, and thus will delete it at some point in the future. If we call this a second time, the returned pnode will have ref=2, and thus the node will never be deleted.

This is independent from the AddRef/Release calls we do here on our own, as it does not change the end result of the ref counter. This means that this PR would not change/fix the ref leaking.

I think this was introduced with the network refactoring, which missed the commit https://github.com/bitcoin/bitcoin/commit/236618061a445d2cb11e722cfac5fdae5be26abb. I remember that you had a discussion somewhere with @OlegGirko about omitting this change, probably due to Dash using ConnectNode in multiple places directly.

The solution would be to backport the missing commit and fix all places where Dash uses ConnectNode to use OpenNetworkConnection+ForNode. I can do this later today or tomorrow, assuming that what I see here is correct :)

Reply:
This shouldn't be the case - `AddRef` is only called if we "connect" to existing node as a masternode now and it wasn't a masternode before, or if we connect to a brand new node as a masternode.

Reply:
Ahhh...I need some coffee as well :)
Ok, postponing the refactoring then and making it part of my 0.13+0.14 backports.